### ISSUE TITLE
HTML Report wrong steps/substeps nesting

### ISSUE THREAD

**Issue by eramahatra** [Codeception/Codeception#4413]
#### What are you trying to achieve?
When using webdrive, I want my steps to be repeated in my HTML report anytime i run the test.
#### What do you get instead?
The html report instead shows only my step once, even if i repeat it in the same suite.
So on for the substeps which then are nested in the wrong steps !

> Provide console output if related. Use `-vvv` mode for more details.

```bash
# codecept run acceptance --html
```
> Provide test source code if related

```php
$I->amDoingSomething($a);
$I->amDoingAnotherThing($b);
$I->comment("this line appears at end of report but it should be in the middle");
$I->amDoingSomething($c);
$I->amDoingAnotherThing($d);
```
### Details

* Codeception version: 
* PHP Version: 2.3.4
* Operating System: ubuntu
* Installation type: Phar
* List of installed packages (`composer show`)
* Suite configuration:

```yml
# Codeception Test Suite Configuration

# suite for acceptance tests.
# perform tests in browser using the WebDriver or PhpBrowser.
# If you need both WebDriver and PHPBrowser tests - create a separate suite.

class_name: AcceptanceTester
modules:
    enabled:
        - AcceptanceHelper
        - Asserts
        - WebDriver:
            port: 4444
            browser: chrome
            url:
            host: localhost
```

**Comment by msgeiger** [comment]
I experienced the same behavior using Step and Page objects and as in the example below, at first glance on the html report, it appears the last step to check for errors is missing.

As mentioned in the post above, the html report i.e. _output/report.html displays the first instance of the Page object method and the steps for that method nested, when if it is used sequentially or multiple times in the test. 

```
namespace Page\Acceptance;

class DetectErrors
{
    public function donSeeErrors()
    {
        $I->dontSeeInPageSource('500 Server')
        ....
    }

namespace AcmeTestSuite\Admin

class Login
{
    public function login(
         \Page\Acceptance\DetectErrors $detect, 
         \Page\Acceptance\Login $login,
         \AcceptenceTester $I
    )
    {
          $I->amOnPage('/login');
          $I->waitForElementVisible('#loginForm', 60);
          $detect->dontSeeErrors();
          $login->fillInSubmitForm();
          $l->waitForElementVisible('#landingPage', 60);

          // The following step does not show in the sequence of steps on the html report. 
          $detect->dontSeeErrors();
     }
}
```

**Comment by reinholdfuereder** [comment]
This is still the rather sad case in Codeception v2.5.4

**Comment by reinholdfuereder** [comment]
Presumably related to or caused by: https://github.com/Codeception/Codeception/pull/3511 and https://github.com/Codeception/Codeception/issues/3410

- There is (at least) one missing case:
```
public function demo(FunctionalTester $I)
{
  $I->demoAction1();
  $I->demoAction1();
}
```
- I think the problem is in https://github.com/Codeception/phpunit-wrapper/blob/ab04a956264291505ea84998f43cf91639b4575d/src/ResultPrinter/HTML.php#L99-L119

**Comment by reinholdfuereder** [comment]
In https://github.com/Codeception/Codeception/issues/3410#issue-169539738 @niclopez lists 6 (test) cases:
* Ideally there would be unit tests for them, but I have unfortunately also continued like this and thus manually tested things
* I tried to recreate code like in his (test) cases (but omitting some proprietary PageObject environment/infrastructure whatever):
  1. The CEST class: "CodeceptionIssue4413Cest"
```
<?php
declare(strict_types=1);

namespace Acceptance\WebDriver\Playground;

/**
 * @group playground
 */
class CodeceptionIssue4413Cest {

  public function case1_ORIG(\AcceptanceTesterWebDriver $I)
  {
    $I->comment('no metaStep');
    $I->comment('no metaStep');
  }

  public function case1(\Codeception\PageObjects\Playground\DemoPageObject $pageObject)
  {
    $pageObject->getAcceptanceTester()->comment('no metaStep');
    $pageObject->getAcceptanceTester()->comment('no metaStep');
  }

  public function case2(\Codeception\PageObjects\Playground\DemoPageObject $pageObject)
  {
    $pageObject->getAcceptanceTester()->comment('no metaStep');
    $pageObject->demoAction1();
    $pageObject->getAcceptanceTester()->comment('no metaStep');
  }

  public function case3(\Codeception\PageObjects\Playground\DemoPageObject $pageObject)
  {
    $pageObject->getAcceptanceTester()->comment('no metaStep');
    $pageObject->demoAction1();
    $pageObject->getAcceptanceTester()->comment('no metaStep');
    $pageObject->demoAction2();
  }

  public function case4(\Codeception\PageObjects\Playground\DemoPageObject $pageObject)
  {
    $pageObject->demoAction1();
    $pageObject->demoAction2();
  }

  public function case5(\Codeception\PageObjects\Playground\DemoPageObject $pageObject)
  {
    $pageObject->getAcceptanceTester()->comment('no metaStep');
    $pageObject->demoAction1WithNestedNoMetastep();
    $pageObject->getAcceptanceTester()->comment('no metaStep');
    $pageObject->demoAction2();
  }

  public function case5b(\Codeception\PageObjects\Playground\DemoPageObject $pageObject)
  {
    $pageObject->getAcceptanceTester()->comment('no metaStep');
    $pageObject->demoAction1WithNestedNoMetastep2();
    $pageObject->getAcceptanceTester()->comment('no metaStep');
    $pageObject->demoAction2();
  }

  public function case6(\Codeception\PageObjects\Playground\DemoPageObject $pageObject)
  {
    $pageObject->demoAction1WithNestedNoMetastep();
    $pageObject->demoAction2();
  }

  public function case6b(\Codeception\PageObjects\Playground\DemoPageObject $pageObject)
  {
    $pageObject->demoAction1WithNestedNoMetastep2();
    $pageObject->demoAction2();
  }

  public function case7(\Codeception\PageObjects\Playground\DemoPageObject $pageObject)
  {
    $pageObject->demoAction1();
    $pageObject->demoAction1();
  }

  public function case8(\Codeception\PageObjects\Playground\DemoPageObject $pageObject)
  {
    $pageObject->demoAction1();
    $pageObject->demoAction1();
    $pageObject->demoAction2();
  }

  public function case9(\Codeception\PageObjects\Playground\DemoPageObject $pageObject)
  {
    $pageObject->demoAction1();
    $pageObject->demoAction2();
    $pageObject->demoAction1();
  }

  public function case10(\Codeception\PageObjects\Playground\DemoPageObject $pageObject)
  {
    $pageObject->demoAction2();
    $pageObject->demoAction1();
    $pageObject->demoAction1();
  }
}
```
  2. The PageObject class "DemoPageObject" used in CEST; but I am not sure if `demoAction1WithNestedNoMetastep()` should be really implemented like this:
```
<?php
declare(strict_types=1);

namespace Codeception\PageObjects\Playground;

class DemoPageObject extends \Codeception\PageObjects\AbstractPageObject {

  /**
   * Prüft ob eh gerade die richtige Seite (d.h. die zur konkreten PageObject Instanz passende, durch den PhpBrowser
   * bzw. WebDriver) angezeigt wird. Muss von jeder konkreten <tt>PageObject</tt> Klasse implementiert werden!
   *
   * @return \Redx\Codeception\PageObjects\AbstractPageObject
   */
  protected function seeCorrectPage(): \Redx\Codeception\PageObjects\AbstractPageObject {
    // TODO: Implement seeCorrectPage() method.
    return $this;
  }

  public function demoAction1(): self {
    $this->getAcceptanceTester()->amOnPage('https://www.google.at');
    $this->getAcceptanceTester()->amOnPage('https://www.google.com');
    return $this;
  }

  public function demoAction2(): self {
    $this->getAcceptanceTester()->amOnPage('https://www.google.de');
    return $this;
  }

  public function demoAction1WithNestedNoMetastep(): self {
    $this->demoAction1();
    $this->getAcceptanceTester()->comment('no metaStep inside a method');
    return $this;
  }

  public function demoAction1WithNestedNoMetastep2(): self {
    $this->demoAction1();
    $this->internalNoMetastep();
    return $this;
  }
  private function internalNoMetastep(): self {
    $this->getAcceptanceTester()->comment('no metaStep inside a method');
    return $this;
  }
}
```
* Results (current situation) with the 6 known (test) cases from https://github.com/Codeception/Codeception/issues/3410
  1. Latest Codeception release 2.5.5 (via PHAR file)
    * Log output: [test.log](https://github.com/Codeception/Codeception/files/3046533/test.log)
    * HTML report for Case 4 looks different then in https://github.com/Codeception/Codeception/issues/3410#issue-169539738, but that was the fix of https://github.com/Codeception/Codeception/issues/3410 anyhow: ![grafik](https://user-images.githubusercontent.com/145182/55606370-d1db1a00-5778-11e9-9e5c-96cf026b0293.png)
    * Log output and HTML report for Case 5 look different then in https://github.com/Codeception/Codeception/issues/3410#issue-169539738, but this may be (as aforementioned) because my implementation of `demoAction1WithNestedNoMetastep()` may not be correct? Nonetheless I think that the actual result is like the expected one (based on my used implementation). And I claim that this was also fixed in https://github.com/Codeception/Codeception/issues/3410
      * Log output:
```
CodeceptionIssue4413Cest: Case5
Signature: Acceptance\WebDriver\Playground\CodeceptionIssue4413Cest:case5
Test: test/codeception/AcceptanceTests.webdriver/Playground/CodeceptionIssue4413Cest.php:case5
Scenario --
 no metaStep
 Codeception\PageObjects\Playground\DemoPageObject: demo action1 with nested no metastep
   I am on page "https://www.google.at"
   I am on page "https://www.google.com"
 no metaStep inside a method
 no metaStep
 Codeception\PageObjects\Playground\DemoPageObject: demo action2
   I am on page "https://www.google.de"
 PASSED
```
      * HTML report: ![grafik](https://user-images.githubusercontent.com/145182/55611337-7f542a80-5785-11e9-84fe-2fe5ae12dbeb.png)
    * Likewise log output and HTML report for Case 6 look different then in https://github.com/Codeception/Codeception/issues/3410#issue-169539738. Again I think that the actual result is like the expected one (based on my used implementation). And I claim that this was also fixed in https://github.com/Codeception/Codeception/issues/3410
      * Log output:
```
CodeceptionIssue4413Cest: Case6
Signature: Acceptance\WebDriver\Playground\CodeceptionIssue4413Cest:case6
Test: test/codeception/AcceptanceTests.webdriver/Playground/CodeceptionIssue4413Cest.php:case6
Scenario --
 Codeception\PageObjects\Playground\DemoPageObject: demo action1 with nested no metastep
   I am on page "https://www.google.at"
   I am on page "https://www.google.com"
 no metaStep inside a method
 Codeception\PageObjects\Playground\DemoPageObject: demo action2
   I am on page "https://www.google.de"
 PASSED
```
      * HTML report: ![grafik](https://user-images.githubusercontent.com/145182/55611358-909d3700-5785-11e9-86eb-8d1840d1180b.png)

  2. Older Codeception release 2.5.1 (via PHAR file): Exactly like in Codeception 2.5.5, i.e. as expected

So this was the current situation with Codeception 2.5.1 and 2.5.5 with the 6 known (test) cases from https://github.com/Codeception/Codeception/issues/3410

**Comment by reinholdfuereder** [comment]
**And now for the (test) cases with Codeception 2.5.5**

* Cases 7 - 10:
  * HTML report with bugs: 
![grafik](https://user-images.githubusercontent.com/145182/55612985-d52ad180-5789-11e9-95d4-4ba2c3bfe347.png)
  * Log output:
    CodeceptionIssue4413Cest: Case7
    Signature: Acceptance\WebDriver\Playground\CodeceptionIssue4413Cest:case7
    Test: test/codeception/AcceptanceTests.webdriver/Playground/CodeceptionIssue4413Cest.php:case7
    Scenario --
     Codeception\PageObjects\Playground\DemoPageObject: demo action1
       I am on page "https://www.google.at"
       I am on page "https://www.google.com"
     Codeception\PageObjects\Playground\DemoPageObject: demo action1
       I am on page "https://www.google.at"
       I am on page "https://www.google.com"
     PASSED

    CodeceptionIssue4413Cest: Case8
    Signature: Acceptance\WebDriver\Playground\CodeceptionIssue4413Cest:case8
    Test: test/codeception/AcceptanceTests.webdriver/Playground/CodeceptionIssue4413Cest.php:case8
    Scenario --
     Codeception\PageObjects\Playground\DemoPageObject: demo action1
       I am on page "https://www.google.at"
       I am on page "https://www.google.com"
     Codeception\PageObjects\Playground\DemoPageObject: demo action1
       I am on page "https://www.google.at"
       I am on page "https://www.google.com"
     Codeception\PageObjects\Playground\DemoPageObject: demo action2
       I am on page "https://www.google.de"
     PASSED

    CodeceptionIssue4413Cest: Case9
    Signature: Acceptance\WebDriver\Playground\CodeceptionIssue4413Cest:case9
    Test: test/codeception/AcceptanceTests.webdriver/Playground/CodeceptionIssue4413Cest.php:case9
    Scenario --
     Codeception\PageObjects\Playground\DemoPageObject: demo action1
       I am on page "https://www.google.at"
       I am on page "https://www.google.com"
     Codeception\PageObjects\Playground\DemoPageObject: demo action2
       I am on page "https://www.google.de"
     Codeception\PageObjects\Playground\DemoPageObject: demo action1
       I am on page "https://www.google.at"
       I am on page "https://www.google.com"
     PASSED

    CodeceptionIssue4413Cest: Case10
    Signature: Acceptance\WebDriver\Playground\CodeceptionIssue4413Cest:case10
    Test: test/codeception/AcceptanceTests.webdriver/Playground/CodeceptionIssue4413Cest.php:case10
    Scenario --
     Codeception\PageObjects\Playground\DemoPageObject: demo action2
       I am on page "https://www.google.de"
     Codeception\PageObjects\Playground\DemoPageObject: demo action1
       I am on page "https://www.google.at"
       I am on page "https://www.google.com"
     Codeception\PageObjects\Playground\DemoPageObject: demo action1
       I am on page "https://www.google.at"
       I am on page "https://www.google.com"
     PASSED

**Comment by reinholdfuereder** [comment]
With my locally patched "codeception/phpunit-wrapper" (PR will follow) the buggy cases 7 - 10 are IMHO fixed:
* HTML report: 
![grafik](https://user-images.githubusercontent.com/145182/55614196-c1cd3580-578c-11e9-84a1-602974553f24.png)

**Comment by reinholdfuereder** [comment]
As aforementioned I am using the Codeception PHAR release, but "interestingly" the **codecept.phar from version 2.5.5 seems to be based upon an old version of "codeception/phpunit-wrapper"!**
* Because "phpunit-wrapper" in Codeception v2.5.5 codecept.phar release is at least missing:
  * https://github.com/Codeception/phpunit-wrapper/commit/42c8fdbf301ee36f6b58cb154972f9148fdcb7d4#diff-5d3444ddf99f7d8bba5a1021e577735d
  * https://github.com/Codeception/phpunit-wrapper/commit/e3f35f1913ade4ab44080118e198fbc585ff8318#diff-5d3444ddf99f7d8bba5a1021e577735d
* Which version of "codeception/phpunit-wrapper" is it really using?
  * According to "composer.json" it might be "7.0.x"?
* Is it the "master" branch of "codeception/phpunit-wrapper"?
* Shouldn't it be the branch "7.1" (of "codeception/phpunit-wrapper") maybe?
* Or should the "master" branch of "codeception/phpunit-wrapper" have some of the changes of branch "7.1" (of "codeception/phpunit-wrapper")?
* Against which version of "codeception/phpunit-wrapper" should I actually prepare the PR?

**Comment by reinholdfuereder** [comment]
This is the PR: https://github.com/Codeception/phpunit-wrapper/pull/62

**Comment by Naktibalda** [comment]
Fixed in phpunit-wrapper version: 8.0.5, 7.7.2, 6.7.1 and 6.1.2.
If you use Codeception 2.4 or newer, just run `composer update` and you will get the patch.
